references/depth/stereo/cascade_evaluation.py

```diff
@@ -0,0 +1,234 @@
+        right_image_pyramid += F.interpolate(image_right, size=ds_shape, mode="bilinear", align_corners=True).unsqueeze(0)
+        
+    flow_init = None
+    for left_image, right_image in zip(reversed(left_image_pyramid), reversed(right_image_pyramid)):
+        flow_pred = model(left_image, right_image, flow_init, iterations=iterations)
+        # flow pred is a list
+        flow_init = flow_pred[-1]
+        
+    return flow_init
+
+@torch.inference_mode()
+def _evaluate(
+    model, args, val_loader, *, padder_mode, print_freq=10, writter=None, step=None, iterations=10, cascades=1, batch_size=None, header=None, save_images=False, save_path="",
+):
+    """Helper function to compute various metrics (epe, etc.) for a model on a given dataset.
+    We process as many samples as possible with ddp.
+    """
+    model.eval()
+    header = header or "Test:"
+    device = torch.device(args.device)
+    metric_logger = utils.MetricLogger(delimiter="  ")
+    
+    iterations = iterations or args.recurrent_updates
+
+    logger = utils.MetricLogger()
+    for meter_name in args.metrics:
+        logger.add_meter(meter_name, fmt="{global_avg:.4f}")
+    if "f1" not in args.metrics:
+        logger.add_meter("f1", fmt="{global_avg:.4f}")
+
+    num_processed_samples = 0
+    with torch.cuda.amp.autocast(enabled=args.mixed_precision, dtype=torch.float16):
+        batch_idx = 0
+        for blob in metric_logger.log_every(val_loader, print_freq, header):
+            image_left, image_right, disp_gt, valid_disp_mask = (x.to(device) for x in blob)
+            padder = utils.InputPadder(image_left.shape, mode=padder_mode)
+            image_left, image_right = padder.pad(image_left, image_right)
+
+            disp_pred = cascade_inference(model, image_left, image_right, iterations, cascades)
+            disp_pred = disp_pred[:, :1, :, :]
+            disp_pred = padder.unpad(disp_pred)
+
+            if save_images:
+                if args.distributed:
+                    rank_prefix = args.rank
+                else:
+                    rank_prefix = 0
+                make_prediction_image_side_to_side(disp_pred, disp_gt, valid_disp_mask, args.save_path, prefix=f"batch_{rank_prefix}_{batch_idx}")
+                
+            metrics, _ = utils.compute_metrics(disp_pred, disp_gt, valid_disp_mask, metrics=logger.meters.keys())
+            num_processed_samples += image_left.shape[0]
+            for name in metrics:
+                logger.meters[name].update(metrics[name], n=1)
+                
+            batch_idx += 1
+    
+    num_processed_samples = utils.reduce_across_processes(num_processed_samples) / args.world_size 
+    
+    print("Num_processed_samples: ", num_processed_samples)
+    if (
+        hasattr(val_loader.dataset, "__len__")
+        and len(val_loader.dataset) != num_processed_samples
+        and torch.distributed.get_rank() == 0
+    ):
+        warnings.warn(
```

Comment:
I've discussed with Yosua and have chosen to follow the reference in the classification script. If we want to change back to the optical flow route, we could do that.

Reply:
We can leave it as it then.

Reply:
Ah yes, just to add context that this is a design decision between `drop_last=True` where we remove some of the eval data vs `drop_last=False` where in this case the data loader will add indices to make sure the data is divisible by the number of GPU and cause duplicate data.

Both case are not ideal in distributed setting (one have duplicate data, one have missing data). But personally I prefer to have duplicate data in this case since the code is simpler (like the classification example). And I think this is more like design choice since both option is not ideal.